src/Session/Handler/Factory/AbstractFactory.php

```diff
@@ -0,0 +1,121 @@
+<?php
+
+namespace Bolt\Session\Handler\Factory;
+
+use Bolt\Helpers\Deprecated;
+use Bolt\Session\OptionsBag;
+use GuzzleHttp\Psr7\Uri;
+use Psr\Http\Message\UriInterface;
+
+/**
+ * Abstract Factory with common functions to help create connections to services.
+ *
+ * @author Carson Full <carsonfull@gmail.com>
+ */
+abstract class AbstractFactory
+{
+    /**
+     * Parse the session options to connection parameters.
+     *
+     * @param OptionsBag $sessionOptions
+     *
+     * @return OptionsBag[]
+     */
+    public function parse(OptionsBag $sessionOptions)
+    {
+        if ($sessionOptions['save_path']) {
+            return $this->parseConnectionsFromSavePath($sessionOptions['save_path']);
+        }
+
+        return $this->parseConnectionsFromOptions($sessionOptions);
+    }
+
+    /**
+     * Parses session save_path into a list of connection parameters.
+     *
+     * @param string $savePath
+     *
+     * @return OptionsBag[]
+     */
+    protected function parseConnectionsFromSavePath($savePath)
+    {
+        $connections = (array) explode(',', $savePath);
+        $connections = array_map('trim', $connections);
+
+        return array_map([$this, 'parseConnectionItemFromSavePath'], $connections);
+    }
+
+    /**
+     * @param string $path
+     *
+     * @return OptionsBag
+     */
+    abstract protected function parseConnectionItemFromSavePath($path);
+
+    /**
+     * Parses session options with array configuration into a list of connection parameters.
+     *
+     * @param OptionsBag $options
+     *
+     * @return OptionsBag[]
+     */
+    protected function parseConnectionsFromOptions(OptionsBag $options)
+    {
+        if ($options->has('connections')) {
+            $connections = $options->get('connections');
+        } elseif ($options->has('connection')) {
+            $connections = [$options->get('connection')];
+        } elseif ($options->has('host') || $options->has('port')) {
+            Deprecated::warn('Specifying "host" and other options directly in session config', 3.3, 'Move them under the "connection" key.');
+
+            $connections = [$options->all()];
+        } else {
+            $connections = [[]];
+        }
+
+        return array_map([$this, 'parseConnectionItemFromOptions'], $connections);
+    }
+
+    /**
+     * @param array|string $item
+     *
+     * @return OptionsBag
+     */
+    abstract protected function parseConnectionItemFromOptions($item);
+
+    /**
+     * @param string $uri
+     *
+     * @return UriInterface
+     */
+    protected function parseUri($uri)
+    {
+        // parse_url() doesn't like "unix:///foo", but parses "unix:/foo" fine.
+        if ($uri && stripos($uri, 'unix://') === 0) {
+            $uri = str_ireplace('unix://', 'unix:', $uri);
+        }
```

Comment:
It's the split on the `} else {` … I ended up just "letting go", but one option would be to put the `else` specific comments in the logic block …